File: src/main/factory/trafficSource/createTrafficSourceFactory.ts
import { CreateTrafficSourceUseCase } from "../../../app/useCases/trafficSource/createTrafficSourceUseCase";
import { CreateTrafficSourceController } from "../../../infra/controllers/trafficSource/createTrafficSourceController";
import { UserGateway } from "../../../infra/gateways/user";
import { PrismaTrafficSourceRepository } from "../../../infra/repositories/trafficSource";

const prismaTrafficSourceRepository = new PrismaTrafficSourceRepository();
const userGateway = new UserGateway();

const createTrafficSourceUseCase = new CreateTrafficSourceUseCase(
  prismaTrafficSourceRepository,
  userGateway
);

const createTrafficSourceController = new CreateTrafficSourceController(
  createTrafficSourceUseCase
);

const createTrafficSource = {
  handle: createTrafficSourceController.handle.bind(
    createTrafficSourceController
  ),
};

export { createTrafficSource };

File: internal/main/factory/trafficSource/createTrafficSourceFactory.go
package trafficSourceFactories

import (
	trafficSourceUseCases "github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/app/useCases/trafficSource"
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/repositories"
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/infra/adapters"
	trafficSourceControllers "github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/infra/controllers/trafficSource"
	contracts "github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/main/contracts"
)

func NewCreateTrafficSourceFactory(trafficSourceRepository repositories.TrafficSourceRepository) contracts.Controller {
	createTrafficSourceUseCase := trafficSourceUseCases.NewCreateTrafficSourceUseCase(
		trafficSourceRepository,
		&adapters.DateGenerator{},
		&adapters.UuidGenerator{},
	)

	return trafficSourceControllers.NewCreateTrafficSourceController(createTrafficSourceUseCase)
}
